File: matlab/examples/waveguide/Coax.m
%
% EXAMPLE / waveguide / coaxial cable
%
% This example demonstrates how to:
%  - setup a coaxial waveguide
%  - use analytic functions for waveguide excitations and voltage/current
%  calculations
% 
%
% Tested with
%  - Matlab 2009b
%  - openEMS v0.0.17
%
% (C) 2010 Thorsten Liebig <thorsten.liebig@uni-due.de>

close all
clear
clc

%% switches & options...
postprocessing_only = 0;
use_pml = 0;            % use pml boundaries instead of mur
openEMS_opts = '';
% openEMS_opts = [openEMS_opts ' --disable-dumps'];

%% setup the simulation %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
numTS = 5000;           %number of timesteps
length = 1000;          %length of the waveguide
unit = 1e-3;            %drawing unit used
coax_rad_i  = 100;      %inner radius
coax_rad_ai = 230;      %inner radius of outer cladding
coax_rad_aa = 240;      %outer radius of outer cladding
mesh_res = [5 5 5];     %desired mesh resolution

physical_constants;

%excitation
f0 = 0.5e9;
epsR = 1;

%% create sim path %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
Sim_Path = 'tmp';
Sim_CSX = 'coax.xml';

if (postprocessing_only==0)
    [status, message, messageid] = rmdir(Sim_Path,'s');
    [status, message, messageid] = mkdir(Sim_Path);
end

%% setup FDTD parameter & excitation function %%%%%%%%%%%%%%%%%%%%%%%%%%%%%
FDTD = InitFDTD(numTS,1e-5);
FDTD = SetGaussExcite(FDTD,f0,f0);
BC = {'PEC','PEC','PEC','PEC','MUR','MUR'};
if (use_pml>0)
    BC = {'PEC','PEC','PEC','PEC','PML_8','PML_8'};
end
FDTD = SetBoundaryCond(FDTD,BC);

%% setup CSXCAD geometry & mesh %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
CSX = InitCSX();
mesh.x = -coax_rad_aa : mesh_res(1) : coax_rad_aa;
mesh.y = mesh.x;
mesh.z = SmoothMeshLines([0 length], mesh_res(3));
CSX = DefineRectGrid(CSX, unit, mesh);

%%% coax
CSX = AddMetal(CSX,'copper');
start = [0,0,0];
stop  = [0,0,length/2];
[CSX,port{1}] = AddCoaxialPort( CSX, 10, 1, 'copper', '', start, stop, 'z', coax_rad_i, coax_rad_ai, coax_rad_aa, 'ExciteAmp', 1,'FeedShift', 10*mesh_res(1) );

start = [0,0,length/2];
stop  = [0,0,length];
[CSX,port{2}] = AddCoaxialPort( CSX, 10, 2, 'copper', '', start, stop, 'z', coax_rad_i, coax_rad_ai, coax_rad_aa );
 
%% define dump boxes... %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
CSX = AddDump(CSX,'Et_','DumpMode',2);
start = [mesh.x(1) , 0 , mesh.z(1)];
stop = [mesh.x(end) , 0 , mesh.z(end)];
CSX = AddBox(CSX,'Et_',0 , start,stop);

CSX = AddDump(CSX,'Ht_','DumpType',1,'DumpMode',2);
CSX = AddBox(CSX,'Ht_',0,start,stop);

%% Write openEMS
if (postprocessing_only==0)
    WriteOpenEMS([Sim_Path '/' Sim_CSX],FDTD,CSX);
    CSXGeomPlot([Sim_Path '/' Sim_CSX]);
    RunOpenEMS(Sim_Path, Sim_CSX, openEMS_opts);
end

%%
freq = linspace(0,2*f0,201);
port = calcPort(port, Sim_Path, freq);

%% plot s-parameter
figure
s11 = port{1}.uf.ref./port{1}.uf.inc;
s21 = port{2}.uf.inc./port{1}.uf.inc;
plot(freq,20*log10(abs(s11)),'Linewidth',2);
hold on
grid on
plot(freq,20*log10(abs(s21)),'r--','Linewidth',2);
xlim([freq(1) freq(end)]);
xlabel('frequency (Hz)')
ylabel('s-para (dB)');

%% plot line-impedance comparison
figure()
ZL_a = ones(size(freq))*Z0/2/pi/sqrt(epsR)*log(coax_rad_ai/coax_rad_i); %analytic line-impedance of a coax
ZL = port{2}.uf.tot./port{2}.if.tot;
plot(freq,real(port{1}.ZL),'Linewidth',2);
xlim([freq(1) freq(end)]);
xlabel('frequency (Hz)')
ylabel('line-impedance (\Omega)');
grid on;
hold on;
plot(freq,imag(port{1}.ZL),'r--','Linewidth',2);
plot(freq,ZL_a,'g-.','Linewidth',2);
legend('\Re\{ZL\}','\Im\{ZL\}','ZL-analytic','Location','Best');

File: python/Tests/Coax.py
# -*- coding: utf-8 -*-
"""
 Coaxial Line Test

 Verifies CoaxialPort setup, S-parameter extraction, and line impedance
 against the analytic formula for a 50 Ω air-filled coaxial line.

 Pass criteria:
   max(dB(S11)) < -0.1 dB   (adequate return loss over band)
   min(dB(S21)) > -50 dB    (cable transmits signal)
   Z_ref within 5 % of analytic value (mid-band)

 Tested with
  - python 3.10
  - openEMS v0.0.35+

 (c) 2026 Thorsten Liebig <thorsten.liebig@gmx.de>

"""

import os, tempfile
import numpy as np

from CSXCAD  import ContinuousStructure
from openEMS import openEMS
from openEMS.physical_constants import *
from openEMS.ports import CoaxialPort

### Setup the simulation
Sim_Path = os.path.join(tempfile.gettempdir(), 'Coax')

unit = 1e-3          # drawing unit in mm
length       = 150   # cable length (mm)
coax_rad_i   = 5     # inner conductor radius  (5 mm)
coax_rad_ai  = 11.5  # inner radius of outer conductor  (ratio 2.3 → 50 Ω)
coax_rad_aa  = 12    # outer radius of outer conductor  (12 mm)
mesh_res     = [0.25, 0.25, 2]  # 0.25 mm transverse, 2 mm axial

f_stop = 3e9
epsR   = 1

### Setup FDTD parameters & excitation
FDTD = openEMS(EndCriteria=1e-4)
FDTD.SetGaussExcite(0, f_stop)
FDTD.SetBoundaryCond(['PEC', 'PEC', 'PEC', 'PEC', 'PML_8', 'PML_8'])

### Setup CSXCAD geometry & mesh
CSX = ContinuousStructure()
FDTD.SetCSX(CSX)
mesh = CSX.GetGrid()
mesh.SetDeltaUnit(unit)

mesh.AddLine('x', np.arange(-coax_rad_aa, coax_rad_aa + mesh_res[0], mesh_res[0]))
mesh.AddLine('y', np.arange(-coax_rad_aa, coax_rad_aa + mesh_res[1], mesh_res[1]))
mesh.AddLine('z', np.arange(0, length + mesh_res[2],                 mesh_res[2]))

### Coaxial ports
copper = CSX.AddMetal('copper')
ports  = []

start = [0, 0, 0]
stop  = [0, 0, length/2]
ports.append(CoaxialPort(CSX, 1, copper, None, start, stop, 'z',
                         coax_rad_i, coax_rad_ai, coax_rad_aa,
                         excite=1, FeedShift=10*mesh_res[2], priority=10))

start = [0, 0, length]
stop  = [0, 0, length/2]
ports.append(CoaxialPort(CSX, 2, copper, None, start, stop, 'z',
                         coax_rad_i, coax_rad_ai, coax_rad_aa, priority=10))

### Run the simulation
FDTD.Run(Sim_Path, cleanup=True, exact_endcriteria=True)

### Post-processing
freq = np.linspace(1e6, f_stop, 201)
for port in ports:
    port.CalcPort(Sim_Path, freq)

s11 = ports[0].uf_ref / ports[0].uf_inc
s21 = ports[1].uf_ref / ports[0].uf_inc

s11_dB = 20 * np.log10(np.abs(s11))
s21_dB = 20 * np.log10(np.abs(s21))

### Pass / fail checks (skip the DC-vicinity bin)
mask = freq > 100e6
print(f'max(dB(S11)) = {np.max(s11_dB[mask]):.1f} dB')
print(f'min(dB(S21)) = {np.min(s21_dB[mask]):.1f} dB,  max(dB(S21)) = {np.max(s21_dB[mask]):.2f} dB')

assert np.max(s11_dB[mask]) < -50, \
    f'FAIL: max(dB(S11)) = {np.max(s11_dB[mask]):.1f} dB, expected < -50 dB'
assert np.min(s21_dB[mask]) > -0.1, \
    f'FAIL: min(dB(S21)) = {np.min(s21_dB[mask]):.1f} dB, expected > -0.1 dB'
assert np.max(s21_dB[mask]) < 0.05, \
    f'FAIL: max(dB(S21)) = {np.max(s21_dB[mask]):.2f} dB, expected < +0.05 dB (sign error?)'

ZL_a   = Z0 / (2*np.pi) / np.sqrt(epsR) * np.log(coax_rad_ai / coax_rad_i)
ZL_num = np.real(ports[0].Z_ref[mask])
rel_err = np.max(np.abs(ZL_num - ZL_a) / ZL_a)
assert rel_err < 0.05, \
    f'FAIL: line impedance error {rel_err*100:.1f} %, expected < 5 %'

print('PASS')

if 0:  # set to 1 for debugging plots
    import matplotlib.pyplot as plt

    fig, axis = plt.subplots(num='S-Parameters', tight_layout=True)
    axis.plot(freq/1e9, s11_dB, linewidth=2, label='$S_{11}$')
    axis.plot(freq/1e9, s21_dB, 'r--', linewidth=2, label='$S_{21}$')
    axis.grid()
    axis.set_xmargin(0)
    axis.set_xlabel('Frequency (GHz)')
    axis.set_ylabel('S-Parameter (dB)')
    axis.legend()

    fig, axis = plt.subplots(num='Line Impedance', tight_layout=True)
    axis.plot(freq/1e9, np.real(ports[0].Z_ref), linewidth=2, label=r'$\Re\{Z_L\}$')
    axis.plot(freq/1e9, np.imag(ports[0].Z_ref), 'r--', linewidth=2, label=r'$\Im\{Z_L\}$')
    axis.plot(freq/1e9, ZL_a * np.ones_like(freq), 'g-.', linewidth=2, label='$Z_{L,analytic}$')
    axis.grid()
    axis.set_xmargin(0)
    axis.set_xlabel('Frequency (GHz)')
    axis.set_ylabel(r'Line impedance $(\Omega)$')
    axis.legend()

    plt.show()
